### ISSUE TITLE
Rotation in the processed MultiDex dataset

### ISSUE THREAD

**Issue by GeorgeWuzy** [4DVLab/DexGrasp-Anything#21]
Hi, 

Thanks for the awesome work!

Originally, Multidex did not include random object rotations. In the Multidex dataset you processed, objects have rotations. How is the aligned rotation of the hand's q achieved? Could you provide a corresponding code for it?

Thank you very much!

Best,
George

**Comment by ymzhong66** [comment]
Hi George,
Thank you for reaching out! I've sent you an email to better assist you with this.
It's been quite a while since I worked on this, but if I remember correctly, I did implement something related to aligning hand qpose rotations with object transformations.
Regarding your question about how to transform the hand's q to achieve aligned displacement and rotations, I can share the code I used back then. However, please keep in mind that this was done a long time ago, and I'm not entirely certain if it's completely correct or directly applicable to your specific case. You'll likely need to verify and adapt it for your setup.
Here's the code for your reference:


import torch
import numpy as np
import transforms3d

def load_pt_file(file_path):
    data = torch.load(file_path)
    return data

def apply_transformations(data, device):
    new_metadata = []

    for item in data['metadata']:
        # Get qpos and rotation matrix
        qpos = item[0].clone()
        rotation_matrix = item[1].clone()

        # Set qpos[0, 9] and qpos[0, 10] to 0
        angle_1 = qpos[9].item()
        angle_2 = qpos[10].item()
        qpos[9] = 0
        qpos[10] = 0

        # Rotation around y-axis
        rotation_matrix_1 = transforms3d.euler.euler2mat(0, angle_1, 0)
        rotation_matrix_1_torch = torch.tensor(rotation_matrix_1, dtype=torch.float32).to(device)

        # Rotation around x-axis
        rotation_matrix_2 = transforms3d.euler.euler2mat(angle_2, 0, 0)
        rotation_matrix_2_torch = torch.tensor(rotation_matrix_2, dtype=torch.float32).to(device)

        # Matrix multiplication to get combined rotation matrix
        combined_rotation_matrix = np.dot(rotation_matrix_1, rotation_matrix_2)
        combined_rotation_matrix_torch = torch.tensor(combined_rotation_matrix, dtype=torch.float32).to(device)

        # Apply transformation to the first three dimensions of qpos
        translation_vector = torch.tensor([0, -0.01, 0.2130], dtype=torch.float32).to(device)
        vector = torch.tensor([0, 0, 0.034], dtype=torch.float32).to(device)
        rotated_vector = torch.matmul(rotation_matrix_1_torch, vector.unsqueeze(1)).squeeze(1)
        transl = qpos[:3].unsqueeze(0).to(device)
        transl = transl - torch.matmul(combined_rotation_matrix_torch, torch.tensor([0, -0.01, 0.2470], dtype=torch.float32).to(device).unsqueeze(1)).squeeze(1) + translation_vector + rotated_vector
        transl = torch.matmul(combined_rotation_matrix_torch.T, transl.T).T
        qpos[:3] = transl.squeeze(0)

        # Apply transformation to rotation matrix
        new_rotation_matrix = torch.matmul(combined_rotation_matrix_torch.T, rotation_matrix)

        # Build new metadata item
        new_metadata.append((qpos, new_rotation_matrix, item[2], item[3]))

    return new_metadata

def save_new_pt_file(data, new_metadata, file_path):
    new_data = {'info': data['info'], 'metadata': new_metadata}
    torch.save(new_data, file_path)
    print(f"保存新的 pt 文件到: {file_path}")

# Example: use this function to read file and save new file
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
file_path = '/inspurfs/group/mayuexin/zym/diffusion+hand/Scene-Diffuser/MultiDex_UR/shadowhand/shadowhand_downsample.pt'
data = load_pt_file(file_path)

# Apply transformations
new_metadata = apply_transformations(data, device)

# Save as new pt file
new_file_path = '/inspurfs/group/mayuexin/datasets/MultiDex_UR/shadowhand/shadowhand_downsample.pt'
save_new_pt_file(data, new_metadata, new_file_path)
